op-batcher/batcher/espresso.go

```diff
@@ -0,0 +1,163 @@
+package batcher
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	espressoVerification "github.com/EspressoSystems/espresso-sequencer-go/verification"
+
+	espressoCommon "github.com/EspressoSystems/espresso-sequencer-go/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// TODO: Pull out to be re-used in op-node for derivation from Espresso
+type Transaction struct {
+	// Namespace of transaction to be published
+	Namespace uint64
+	// TEE attestation to be verified by op-node
+	TeeAttn []byte
+	// Frames serialized as they would be for posting to L1 as calldata
+	CallData []byte
+}
+
+func (t Transaction) toEspresso() espressoCommon.Transaction {
+	payload := append(t.TeeAttn, t.CallData...)
+	return espressoCommon.Transaction{
+		Namespace: t.Namespace,
+		Payload:   payload,
+	}
+}
+
+func (l *BatchSubmitter) waitForFinality(height uint64, rawHeader json.RawMessage, header *espressoCommon.HeaderImpl) error {
+	timer := time.NewTimer(2 * time.Minute)
+	defer timer.Stop()
+
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	var snapshot espressoCommon.BlockMerkleSnapshot
+
+Loop:
+	for {
+		select {
+		case <-ticker.C:
+			res, err := l.EspressoLightClient.FetchMerkleRoot(height, nil)
+			if err == nil {
+				snapshot = res
+				break Loop
+			}
+		case <-timer.C:
+			return fmt.Errorf("failed to fetch merkle root")
+		}
+	}
+
+	if snapshot.Height <= height {
+		return fmt.Errorf("snapshot height is less than or equal to the requested height")
+	}
+
+	nextHeader, err := l.Espresso.FetchHeaderByHeight(l.shutdownCtx, snapshot.Height)
+	if err != nil {
+		return fmt.Errorf("error fetching the snapshot header (height: %d): %w", snapshot.Height, err)
+	}
+
+	proof, err := l.Espresso.FetchBlockMerkleProof(l.shutdownCtx, snapshot.Height, height)
+	if err != nil {
+		return fmt.Errorf("error fetching merkle proof")
+	}
+
+	blockMerkleTreeRoot := nextHeader.Header.GetBlockMerkleTreeRoot()
+
+	log.Info("Verifying merkle proof", "height", height)
```

Comment:
Here we log `height`, but in most function calls we use `snapshot.Height`. Should we check whether they are the same?

Reply:
Follow-up question: If they are not the same, i.e., `snapshot.Height` is larger (since the smaller case has been checked), what should we do?

I was thinking we should make sure each finalized height is properly handled without skipping. If so, we may need extra handling here.

Reply:
Just to double-check, are these two comments resolved?

Reply:
Oops, sorry, I've somehow managed to ignore this comment twice.
My understanding is that we should be fine as long as returned snapshot height is >= requested height and verification passes, as we're still running verification against our target header.